zio-http/src/main/scala/zhttp/service/server/content/handler/ServerResponseHandler.scala

```diff
@@ -0,0 +1,113 @@
+package zhttp.service.server.content.handler
+
+import io.netty.buffer.{ByteBuf, Unpooled}
+import io.netty.channel.ChannelHandler.Sharable
+import io.netty.channel.{ChannelHandlerContext, SimpleChannelInboundHandler}
+import io.netty.handler.codec.http.HttpResponseStatus.INTERNAL_SERVER_ERROR
+import io.netty.handler.codec.http.HttpVersion.HTTP_1_1
+import io.netty.handler.codec.http._
+import zhttp.core.Util
+import zhttp.http.{HTTP_CHARSET, HttpData, Response, Status}
+import zhttp.service.server.ServerTimeGenerator
+import zhttp.service.{ChannelFuture, HttpRuntime}
+import zio.stream.ZStream
+import zio.{UIO, ZIO}
+
+/**
+ * Transforms an internal Response into a HttpResponse.
+ */
+@Sharable
+private[zhttp] case class ServerResponseHandler[R](runtime: HttpRuntime[R], serverTime: ServerTimeGenerator)
+    extends SimpleChannelInboundHandler[Either[Throwable, Response[R, Throwable]]](false) {
+  override def channelRead0(ctx: ChannelHandlerContext, response: Either[Throwable, Response[R, Throwable]]): Unit = {
+    try {
+      response match {
+        case Left(err)               =>
+          ctx.writeAndFlush(serverErrorResponse(err))
+        case Right(internalResponse) =>
+          internalResponse.status match {
+            case Status.NOT_FOUND =>
+              ctx.writeAndFlush(notFoundResponse)
+            case Status.OK        =>
+              ctx.write(decodeResponse(internalResponse))
+              internalResponse.data match {
+                case HttpData.Empty                =>
+                  ctx.writeAndFlush(LastHttpContent.EMPTY_LAST_CONTENT)
+                case HttpData.Text(data, charSet)  =>
+                  ctx.writeAndFlush(new DefaultLastHttpContent(Unpooled.copiedBuffer(data, charSet)))
+                case HttpData.BinaryByteBuf(data)  =>
+                  ctx.writeAndFlush(new DefaultLastHttpContent(data))
+                case HttpData.BinaryStream(stream) =>
+                  runtime.unsafeRun(ctx) {
+                    writeStreamContent(stream)(ctx)
+                  }
+                case HttpData.BinaryChunk(data)    =>
+                  ctx.write(data)
+              }
+
+            case _ =>
+              ctx.writeAndFlush(decodeResponse(internalResponse))
+          }
+      }
+    } catch {
+      case err: Throwable => ctx.writeAndFlush(serverErrorResponse(err))
+    }
+    ()
+
+  }
+
+  override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable): Unit = {
+    ctx.writeAndFlush(serverErrorResponse(cause)): Unit
```

Comment:
needs unit tests. Also this won't really fix #686. Reason being — Once the connection is closed, you can't write anything more on the channel. That's the reason you got this error in the first place.